### PR TITLE
Added limit for the size of Uploading Images

### LINKED ISSUE #1667: Bug Report: Images Exceeding Size Limits are Being Uploaded
*Describe the bug*
Currently, images exceeding size limits for display are being uploaded, even when an error is given. 

*Expected behavior*
If an image is too large, an error should be given AND the image should NOT be uploaded. 

*Actual behavior*
The image is being uploaded and stored even though an error is seen. 

*Additional details*
PLEASE TAKE NOTE: Please limit the size of all image related mutations to 3 MB. This limit should be included as a parameter in the .env file.  Please also make the setup.ts file prompt for it too when the API setup is done.

Please also note that this is being worked on in Talawa-Admin: https://github.com/PalisadoesFoundation/talawa-admin/issues/1405

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by aarishshahmohsin**
**What kind of change does this PR introduce?**
This PR adds a image upload size limit.

**Issue Number:**

Fixes #1667 

**Did you add tests for your changes?**
Yes

**Snapshots/Videos:**

https://github.com/PalisadoesFoundation/talawa-api/assets/49566965/a68f2f05-7943-4782-9559-a08207a6dc62

**Summary**
This PR adds a image size upload limit as an environment variable set default to 3000 KB. It also adds a setup prompt to se the limit according to the setup requirement with the maximum limit being 3000 KB.

**Does this PR introduce a breaking change?**
No

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes

### CODE REVIEW

File: setup.ts [hunk 1]
```diff
@@ -529,9 +544,20 @@ function validateRecaptcha(string: string): boolean {
...
 }
 
+/**
+ * The function validates whether a given image size is less than 3000 and greater than 0.
+ * @param string - The `number` parameter represents the input size of the string
+ * validated. In this case, it is expected to be a number less than 3000 and greater than 0.
+ * @returns a boolean value.
+ */
+function validateImageFileSize(size: number): boolean {
+  return size > 0 && size <= 3000;
```
Comment (palisadoes) [inline setup.ts:554]: 1. People don't think in kB for images. 
2. Convert this to MB.

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comment

File: setup.ts [hunk 2]
```diff
@@ -996,6 +1022,21 @@ async function main(): Promise<void> {
     }
   }
 
+  const { imageSizeLimit } = await inquirer.prompt([
+    {
+      type: "input",
+      name: "imageSizeLimit",
+      message: "Enter the maximum size limit of Images uploaded (in Kb)",
```
Comment (palisadoes) [inline setup.ts:1029]: 1. People don't think in kB for images. 
2. Convert this to MB.

File: setup.ts [hunk 3]
```diff
@@ -529,6 +544,16 @@ function validateRecaptcha(string: string): boolean {
...
 }
 
+/**
+ * The function validates whether a given image size is less than 3 and greater than 0.
+ * @param string - The `number` parameter represents the input size of the string
+ * validated. In this case, it is expected to be a number less than 3 and greater than 0.
+ * @returns a boolean value.
+ */
+function validateImageFileSize(size: number): boolean {
+  return size > 0 && size <= 3;
```
Comment (palisadoes) [inline setup.ts:554]: Make the max 20. Leave the default at 3

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Add entries for these new parameters in the `.env.sample` file
2. What is the default value if none is defined in the `.env` file? It's not clear
3. Does the script default to the configured value if found? The other parameters do this.

File: .env.sample [hunk 4]
```diff
@@ -72,4 +72,8 @@ LOG_LEVEL = info
 
 REDIS_HOST=
 REDIS_PORT=
-REDIS_PASSWORD=
\ No newline at end of file
+REDIS_PASSWORD=
+
+# this environment variable is for setting the environment variable for Image Upload size
```
Comment (palisadoes) [inline .env.sample:77]: You have not stated the units of measurement. This could be 3000 of anything.

File: setup.ts [hunk 5]
```diff
@@ -529,6 +544,16 @@ function validateRecaptcha(string: string): boolean {
...
 }
 
+/**
+ * The function validates whether a given image size is less than 20 and greater than 0.
+ * @param string - The `number` parameter represents the input size of the string
+ * validated. In this case, it is expected to be a number less than 20 and greater than 0.
+ * @returns a boolean value.
+ */
+function validateImageFileSize(size: number): boolean {
+  return size > 0 && size <= 20;
```
Comment (palisadoes) [inline setup.ts:554]: 1. If it's greater than 20, just set the value to 20
2. Use a variable instead of 20 so that we can easily change it in future

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comments

File: setup.ts [hunk 6]
```diff
@@ -996,6 +1026,20 @@ async function main(): Promise<void> {
     }
   }
 
+  const { imageSizeLimit } = await inquirer.prompt([
+    {
+      type: "input",
+      name: "imageSizeLimit",
+      message:
+        "Enter the maximum size limit of Images uploaded (in MB) max: 20",
```
Comment (palisadoes) [inline setup.ts:1034]: The max in the text should be set by the variable

File: setup.ts [hunk 7]
```diff
@@ -996,6 +1026,20 @@ async function main(): Promise<void> {
...
+      default: 3,
+      validate: (input: number) =>
+        validateImageFileSize(input) || "Enter a valid number between 0 and 20",
```
Comment (palisadoes) [inline setup.ts:1037]: The max in the text should be set by the variable

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comments

**Comment by aarishshahmohsin** [comment]
@SiddheshKukade @palisadoes @xoldyckk Is there any update?

**Comment by aarishshahmohsin** [comment]
@SiddheshKukade @palisadoes @xoldyckk Is there any update?

**Comment by aarishshahmohsin** [comment]
@palisadoes Is there any update with regards to the PR. I am not able to move on to other issues

File: src/utilities/encodedImageStorage/uploadEncodedImage.ts [hunk 8]
```diff
@@ -3,17 +3,37 @@ import * as fs from "fs";
...
 export const uploadEncodedImage = async (
   encodedImageURL: string,
   previousImagePath?: string | null
 ): Promise<string> => {
   const isURLValidImage = encodedImageExtentionCheck(encodedImageURL);
 
+  const data = encodedImageURL.replace(/^data:image\/\w+;base64,/, "");
+  const stringLength = data.length;
+  const sizeInKb =
+    (4 * Math.ceil(stringLength / 3) * 0.5624896334383812) / 1000;
```
Comment (EshaanAgg) [inline src/utilities/encodedImageStorage/uploadEncodedImage.ts:24]: Add comments to explain this calculation. If there are any constants, like `0.5624896334383812`, declare then as a const with the appropriate name, and then use the same.

File: tests/utilities/encodedImageStorage/uploadEncodedImage.spec.ts [hunk 9]
```diff
@@ -39,6 +53,17 @@ describe("src -> utilities -> encodedImageStorage -> uploadEncodedImage", () =>
     }
   });
 
+  it("should not create new image if it is bigger than the limit", async () => {
+    const size = Number(process.env.IMAGE_SIZE_LIMIT_KB);
+    try {
+      const img = "data:image/jpg;base64," + generateRandomString(size + 1000);
+      console.log(img.length);
```
Comment (EshaanAgg) [inline tests/utilities/encodedImageStorage/uploadEncodedImage.spec.ts:60]: Remove the un-necesaary `console.log`

**Review by EshaanAgg — CHANGES_REQUESTED** [review]
Minor code-quality changes are required.

**Review by EshaanAgg — APPROVED** [review]

File: setup.ts [hunk 10]
```diff
@@ -116,6 +117,24 @@ async function accessAndRefreshTokens(
   }
 }
 
+//set the image size upload environment variable
+/**
+ * The function `setImageUploadSize` sets the image upload size environment variable and changes the .env file
+ * @returns The function `checkExistingRedis` returns a void Promise.
+ */
+async function setImageUploadSize(size: number): Promise<void> {
```
Comment (palisadoes) [inline setup.ts:125]: How are negative and zero value numbers going to be handled?

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comment

File: setup.ts (hunk 10 above)
Comment (aarishshahmohsin) [reply setup.ts:125]:
```
const { imageSizeLimit } = await inquirer.prompt([
    {
      type: "input",
      name: "imageSizeLimit",
      message: `Enter the maximum size limit of Images uploaded (in MB) max: ${
        MAXIMUM_IMAGE_SIZE_LIMIT_KB / 1000
      }`,
      default: 3,
      validate: (input: number) =>
        validateImageFileSize(input) ||
        `Enter a valid number between 0 and ${
          MAXIMUM_IMAGE_SIZE_LIMIT_KB / 1000
        }`,
    },
  ]);
```
while taking the input it firstly validates using the "validateImageFileSize" function
```
function validateImageFileSize(size: number): boolean {
  return size > 0;
}
```

**Review by palisadoes — APPROVED** [review]

**Pull request merged**